graylog-storage-opensearch3/src/test/java/org/graylog/storage/opensearch3/testing/client/mock/MockedOpensearchClient.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.storage.opensearch3.testing.client.mock;
+
+import org.graylog.storage.opensearch3.OfficialOpensearchClient;
+import org.opensearch.client.opensearch.OpenSearchAsyncClient;
+import org.opensearch.client.opensearch.OpenSearchClient;
+
+import java.util.LinkedHashSet;
+
+public class MockedOpensearchClient {
```

Comment:
This is a very interesting and fresh concept, with a neat way to use it and some interesting advantages!!!

I have one big concern - this way of testing couples tested class with some implementation details of the OS client, that can potentially change. We assume that certain client methods are based on certain REST calls. When this assumption fails, there is a risk that those type of tests will fail DESPITE the fact that our code still works correctly.

A smaller concern - MockedOpensearchClient is a misleading name. I have troubles suggesting a better one (OpensearchClientWithStubbedTransportCalls ???), but we should find smth better, so that no one is fooled that this is a typical mock for unit tests...